input/pipewire/pipewire.go

```diff
@@ -0,0 +1,228 @@
+	pwObjs, err := pwDump()
+	if err != nil {
+		return nil, err
+	}
+
+	pwSinks := pwObjs.AudioSinks()
+
+	devices := make([]input.Device, len(pwSinks))
+	for i, device := range pwSinks {
+		devices[i] = AudioDevice{device.Info.Props.NodeName}
+	}
+
+	return devices, nil
+}
+
+func (p Backend) DefaultDevice() (input.Device, error) {
+	return AudioDevice{"auto"}, nil
+}
+
+func (p Backend) Start(cfg input.SessionConfig) (input.Session, error) {
+	return NewSession(cfg)
+}
+
+type AudioDevice struct {
+	name string
+}
+
+func (d AudioDevice) String() string {
+	return d.name
+}
+
+type catnipProps struct {
+	ApplicationName string `json:"application.name"`
+	CatnipID        string `json:"catnip.id"`
+}
+
+// Session is a PipeWire session.
+type Session struct {
+	session    execread.Session
+	props      catnipProps
+	targetName string
+}
+
+// NewSession creates a new PipeWire session.
+func NewSession(cfg input.SessionConfig) (*Session, error) {
+	currentProps := catnipProps{
+		ApplicationName: "catnip",
+		CatnipID:        generateID(),
+	}
+
+	propsJSON, err := json.Marshal(currentProps)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal props")
+	}
+
+	dv, ok := cfg.Device.(AudioDevice)
+	if !ok {
+		return nil, fmt.Errorf("invalid device type %T", cfg.Device)
+	}
+
+	target := "0"
+	if dv.name == "auto" {
+		target = dv.name
+	}
+
+	args := []string{
+		"pw-cat",
+		"--record",
+		"--format", "f32",
+		"--rate", fmt.Sprint(cfg.SampleRate),
+		"--latency", fmt.Sprint(cfg.SampleSize),
+		"--channels", fmt.Sprint(cfg.FrameSize),
+		"--target", target, // see .relink comment below
+		"--quality", "0",
+		"--media-category", "Capture",
+		"--media-role", "DSP",
+		"--properties", string(propsJSON),
+		"-",
+	}
+
+	session := Session{
+		session:    *execread.NewSession(args, true, cfg),
+		props:      currentProps,
+		targetName: dv.name,
+	}
+	session.session.OnStart = session.onStart
+
+	return &session, nil
+}
+
+// Start starts the session. It implements input.Session.
+func (s *Session) Start(ctx context.Context, dst [][]input.Sample, kickChan chan bool, mu *sync.Mutex) error {
+	return s.session.Start(ctx, dst, kickChan, mu)
+}
+
+func (s *Session) onStart(ctx context.Context, cmd *exec.Cmd) error {
+	var err error
+	for i := 0; i < 20; i++ {
+		if err = s.relink(ctx, cmd); err == nil {
+			return nil
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	return err
+}
+
+// We do a bit of tomfoolery here. Wireplumber actually is pretty incompetent at
+// handling target.device, so our --target flag is pretty much useless. We have
+// to do the node links ourselves.
+//
+// Relevant issues:
+//
+//   - https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/2731
+//   - https://gitlab.freedesktop.org/pipewire/wireplumber/-/issues/358
+//
+func (s *Session) relink(ctx context.Context, cmd *exec.Cmd) error {
+	// No relinking needed if we're not connecting to a specific device.
+	if s.targetName != "auto" {
```

Comment:
Changing this to `if s.target == "auto" {` fixes the issue with exiting without linking the pipes.